test/jdk/java/util/concurrent/Executors/AutoShutdown.java

```diff
@@ -23,89 +23,115 @@
 
 /*
  * @test
- * @bug 6399443
- * @summary Check for auto-shutdown and gc of singleThreadExecutors
- * @library /test/lib
- * @run main/othervm/timeout=1000 AutoShutdown
- * @author Martin Buchholz
+ * @bug 6399443 8302899
+ * @summary Test that Executors.newSingleThreadExecutor wraps an ExecutorService that
+ *    automatically shuts down and terminates when the wrapper is GC'ed
+ * @modules java.base/java.util.concurrent:+open
+ * @run junit AutoShutdown
  */
 
-import static java.util.concurrent.Executors.defaultThreadFactory;
-import static java.util.concurrent.Executors.newSingleThreadExecutor;
+import java.lang.reflect.Field;
+import java.time.Duration;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
 
-import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.jupiter.params.provider.Arguments;
+import static org.junit.jupiter.api.Assertions.*;
 
-import java.lang.ref.WeakReference;
-import java.util.Arrays;
-import java.util.concurrent.ConcurrentLinkedQueue;
-import java.util.concurrent.CountDownLatch;
-import java.util.concurrent.Executor;
-import java.util.concurrent.TimeUnit;
-import jdk.test.lib.Utils;
+class AutoShutdown {
+
+    private static Stream<Supplier<ExecutorService>> executors() {
+        return Stream.of(
+            () -> Executors.newSingleThreadExecutor(),
+            () -> Executors.newSingleThreadExecutor(Executors.defaultThreadFactory())
+        );
+    }
 
-public class AutoShutdown {
-    static final long LONG_DELAY_MS = Utils.adjustTimeout(10_000);
+    private static Stream<Arguments> executorAndQueuedTaskCounts() {
```

Comment:
I've long wondered about "cartesian products" of Streams, so off to stackoverflow, which (TIL) advises a chain of flatmaps except at the end where you just put a non-flat map.  So better style seems

    private static Stream<Arguments> executorAndQueuedTaskCounts() {
        int[] queuedTaskCounts = { 0, 1, 2 };
        return executors()
            .flatMap(s -> IntStream.of(queuedTaskCounts)
                     .mapToObj(i -> Arguments.of(s, i)));
    }